File: assign1/prodcons2.cpp
#include "queue.h"
#include <iostream> 
#include <thread> 
#include <mutex> 
#include <condition_variable> 
using namespace std; 

mutex mtx;   
condition_variable cv;
int BUFFSIZE = 20;
int NUMITEMS = 100; 
int * conRes;
Queue pcq(BUFFSIZE);

void producer(int n) {
  int core = sched_getcpu();
  cout << "Producer starting on core " + to_string(core) + "\n";

  for(int i = 1; i <= NUMITEMS; ++i){
    unique_lock<mutex> lck(mtx);
    while(pcq.isFull()) 
    {
      cv.notify_all();
      cv.wait(lck);
    } 
    pcq.add(i);
    lck.unlock();
    cv.notify_one();
    cout << "Producer added " + to_string(i) + " (qsz: " + to_string(pcq.size()) + ")\n"; 
  }
  for(int i = 0; i < n; ++i){
    unique_lock<mutex> lck(mtx);
    while(pcq.isFull()) 
    {
      cv.notify_all();
      cv.wait(lck);
    } 
    pcq.add(-1);
    lck.unlock();
  }
  cv.notify_all();
  cout << "Producer ending\n";
}
  
void consumer(int tid) {
  int r, c = 0;
  int core = sched_getcpu();
  cout << "Consumer[" + to_string(tid) + "] starting on core " + to_string(core) + "\n";

  while(r != -1) {
    unique_lock<mutex> lck(mtx);
    while(pcq.isEmpty()) 
    {
      cv.notify_one();
      cv.wait(lck);
    } 
    r = pcq.remove();
    lck.unlock();
    cv.notify_one();

    if (r != -1) {
      ++c;
      cout << "Consumer[" + to_string(tid) + "] removed " + to_string(r) 
        + " (qsz: " + to_string(pcq.size()) + ")\n"; 
    }
  }
  cv.notify_one();
  cout << "Consumer[" + to_string(tid) + "] ending\n";
  conRes[tid] = c;
}

int main(int argc, char **argv) {
  int numCons, tot = 0;
  
  if(argv[1]) numCons = atoi(argv[1]);   //gets the number of
  else numCons = 1;                      //consumers

  conRes = new int [numCons];            //inits an array for to hold 
                                         //the number each consumer consumed
  thread pthd, cthd[numCons];

  cout << "Prodcons2 with "+ to_string(numCons) + " consumers and a single producer\n";
  
  pthd = thread(producer,numCons);       //start producer
  for (int k = 0; k < numCons; k++)      //start consumers
    cthd[k] = thread(consumer, k);

  pthd.join();                           //wait for the producer to finish
  for (int k = 0; k < numCons; k++)      //join consumers
    cthd[k].join();

  cout << "Consumer stats: [";
  for(int i = 0; i < numCons; ++i) {
    cout << to_string(conRes[i]) + ",";
    tot += conRes[i];
  }
  cout << "] = " + to_string(tot) + "\n";

  cout << "Main: all done! Now stop bothering me!\n";
  delete(conRes);
}  

File: assign1/ProdCons2.java
import java.util.LinkedList;
import java.util.Queue;

class ProdCons2 extends Thread {
  static final int BUFSIZE = 20; 
  static final int NUMITEMS = 100; 
  static int ret[];
  static Object lck = new Object();
  static Queue<Integer> q = new LinkedList<>();

  public static void main(String[] args) throws InterruptedException {
    int k = 0;
    int tot = 0;
    if(args.length == 0){ 
      k = 1;
    } else { 
      k = Integer.parseInt(args[0]);
    }
    final Prod p = new Prod();
    final Con[] cons = new Con[k];
    ret = new int [k];
    
    
    Thread prod = new Thread() {
      public void run() 
      {
        try {
          p.producer();
        } 
        catch (InterruptedException e) {
          e.printStackTrace();
        }
      }
    };
    prod.start();

    for(int i = 0; i < k; ++i){
      cons[i] = new Con(i);
    }
    for(int i = 0; i < k; ++i) {
      cons[i].start();
    }

    prod.join();
    for(int i = 0; i < k; ++i) {  
      cons[i].join();
    }

    System.out.print("Consumer Stats: [");
    for(int i = 0; i < k; ++i) {  
      System.out.print(ret[i] + ",");
      tot += ret[i];
    }
    System.out.println("] total = " + tot);
    
        
    System.out.println("Main: all done! So leave me alone!");
  }
  public static class Prod {
    Prod() {}

    private void producer() throws InterruptedException{ 
      int i = 0;
      System.out.println("Producer is starting...");
      while(i < NUMITEMS){
        ++i;
        synchronized(lck) {
          while(q.size() == BUFSIZE) {
            lck.notifyAll();
            lck.wait();
          }
          q.add(i);
          lck.notifyAll();
        }
        System.out.println("Producer added " + i + " (qsz: " + q.size() + ")");
      }
      synchronized(lck) {
        for(i = 0; i < NUMITEMS; ++i)
          q.add(-1);
          lck.notifyAll();
      }
      System.out.println("Producer is ending...");
    }
  }

  public static class Con extends Thread{
    private int nme = 0;
    Con(int tid) {
      nme = tid;
    }

    private int consumer()  throws InterruptedException{ 
      System.out.println("Consumer[" + currentThread().getName() + "] is starting...");
      int num = 0;
      int i = 0;

      while(num != -1){
        synchronized(lck) {
          while(q.size() == 0) {
            lck.wait();
          }
          num = q.remove();
          lck.notify();
        }
        if( num != -1) {
          System.out.println("Consumer[" + currentThread().getName() 
              + "] consumed " + num + " (qsz: " + q.size() + ")");
          ++i;
        }
      }

      System.out.println("Consumer[" + currentThread().getName() + "] is ending...");
      return i;
    }

    public void run(){
      try {
        int i = this.consumer();
        ret[this.nme] = i;

      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }
  }
}

